File: lab6/c/shortest_path.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>

#include "graph.h"
#include "pq.h"
#include "shortest_path.h"
#include "sp_algorithms.h"

sssp_result_t *sssp_result_new(size_t N, node_t src, node_t dst, bool ncyc,node_t *p, weight_t *d, unsigned long long c) {
  sssp_result_t *r = CALLOC(sssp_result_t,1);
  r->N=N;
  r->src=src;
  r->dst=dst;
  r->pred=p;
  r->has_negative_cycle=ncyc;
  r->dist=d;
  r->relax_count=c;
  return r;
}

void sssp_result_delete(sssp_result_t *r) {
  if (r->pred) free(r->pred);
  if (r->dist) free(r->dist);
  free(r);
}

sp_result_t *sp_result_new(node_t src, node_t dst, path_t *p, weight_t d, unsigned long long c) {
  sp_result_t *r = CALLOC(sp_result_t,1);
  r->src=src;
  r->dst=dst;
  r->path=p;
  r->dist=d;
  r->relax_count=c;
  return r;
}

void sp_result_delete(sp_result_t *r) {
  if (r->path) path_delete(r->path);
  free(r);
}


sp_result_t *sssp_to_sp_result(sssp_result_t *a, node_t dst) {
  path_t *p = NULL;

  assert(a->dst==INVALID_NODE || a->dst==dst);

  if (a->pred[dst] != INVALID_NODE) {
    p = path_from_pred(a->pred,dst);
  }

  sp_result_t *r = sp_result_new(a->src, dst, p,a->dist[dst],a->relax_count);

  sssp_result_delete(a);

  return r;
}

void print_path(path_t const *p) {
  if (p == NULL) printf("NULL");
  else if (path_len(p) == 0) printf("EMP");
  else {
    printf("%zu",path_get(p,0));
    for (size_t i=1;i<path_len(p);++i)
      printf(", %zu",path_get(p,i));

  }
}

void print_sp_result(sp_result_t const *r) {
  printf("Distance: "); print_weight_stdout(r->dist); printf("\n");
  printf("Path: "); print_path(r->path); printf("\n");
  printf("# Relaxed nodes: %llu\n", r->relax_count);
}

void print_sssp_result(FILE *f, sssp_result_t const *r) {
  size_t N = r->N;
  size_t NN = N<10?N:10;

  fprintf(f,"Distmap:");
  for (size_t i=0; i<NN; ++i) {
    fprintf(f," "); print_weight(f,r->dist[i]);
  }

  if (N<NN) {
    fprintf(f," ...\n");
  } else {
    fprintf(f,"\n");
  }

}

File: lab6/python/shortest_path.py
from graph import graph_t, path_t, INVALID_NODE
from pq import DPQ_t
# import sp_algorithms

#
 # The main components of a single-source shortest path result are a predecessor map and a distance map (encoded as arrays).
 #
 # The distance map assigns each node to its distance from the source node.
 #    If the node is unreachable, the distance is inf.
 #    If the node is reachable via a negative weight cycle, the distance is -inf.
 #
 # The predecessor map assigns nodes to predecessor nodes on a shortest path or negative cycle.
 #    If the node is unreachable, the predecessor must be INVALID_NODE.
 #    If the node has finite distance, its predecessor must be the previous node on a shortest path. EXCEPTION: The predecessor of the start node is the start node itself.
 #    If the node has -inf distance, following the predecessors must yield a negative-weight cycle. NOTE: In this case, there is no exception for the start node!
 #
 #/
class sssp_result_t:
    def __init__(self, N, src, dst, ncyc, p, d, c):
        self.N = N # Number of nodes, to make result more self contained
        self.src = src # Start node
        self.dst = dst # If != INVALID_NODE, this is a partial result, only accurate for dst
        self.pred = p # Predecessor map
        self.has_negative_cycle = ncyc # True if negative cycle was detected. Pred and dist are NULL then.
        self.dist = d # Distance map
        self.relax_count=c # Number of relaxed edges
        
    # Convert sssp to sp result.
    def sssp_to_sp_result(self, dst):
        p = None
        
        assert(self.dst == INVALID_NODE or self.dst == dst)
        
        if (not self.pred[dst] == INVALID_NODE):
            p = path_t(self.pred,dst)
            
        r = sp_result_t(self.src, dst, p, self.dist[dst], self.relax_count)
        
        return r
        
    def print_sssp_result(self, f):
        N = self.N
        if (N<10):
            NN = N
        else:
            NN = 10
       
        f.write("Distmap:")
        for i in range(0, NN):
            f.write(" ")
            self.dist[i].print_weight(f)
        
        if (N < NN):
            f.write("...\n")
        else:
            f.write("\n")
        
        
##
 # Structure describing the result of a shortest path search between two nodes
 #/        
class sp_result_t:
    def __init__(self, src, dst, p, d, c):
        self.src = src                      # Start node
        self.dst = dst                       # Target node
        self.path = p                     # Result path, None if unreachable
        self.dist = d                   # Distance, INF if unreachable
        self.relax_count = c   # Number of relaxed edges
        
    def print_sp_result(self, f):
        f.write("Distance: ")
        self.dist.print_weight(f)
        f.write("\n")
        
        f.write("Path: ")
        print_path(f, self.path)
        f.write("\n")
        
        f.write("# Relaxed nodes: %d\n" % (self.relax_count))
        
def print_path(f, p):
    if (p == None):
        f.write("NULL")
    elif (p.path_len() == 0):
        f.write("EMP")
    else:
        f.write("%d " % (p.path_get(0)))
        for i in range(1, p.path_len()):
            f.write("%d " % (p.path_get(i)))
            

        

